File: examples/go/agg-hand/handlers/state.go
// Package handlers implements hand aggregate command handlers.
package handlers

import (
	"encoding/hex"
	"strings"

	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
)

// HandState represents the current state of a hand aggregate.
type HandState struct {
	HandID      string
	TableRoot   []byte
	HandNumber  int64
	GameVariant examples.GameVariant

	// Deck state
	RemainingDeck []*examples.Card

	// Player state
	Players map[string]*PlayerHandState // player_root_hex -> state

	// Community cards
	CommunityCards []*examples.Card

	// Betting state
	CurrentPhase     examples.BettingPhase
	ActionOnPosition int32
	CurrentBet       int64
	MinRaise         int64
	Pots             []*PotState

	// Positions
	DealerPosition     int32
	SmallBlindPosition int32
	BigBlindPosition   int32

	Status string // "dealing", "betting", "showdown", "complete"
}

// PlayerHandState represents a player's state in the hand.
type PlayerHandState struct {
	PlayerRoot    []byte
	Position      int32
	HoleCards     []*examples.Card
	Stack         int64
	BetThisRound  int64
	TotalInvested int64
	HasActed      bool
	HasFolded     bool
	IsAllIn       bool
}

// PotState represents a pot (main or side).
type PotState struct {
	Amount          int64
	EligiblePlayers [][]byte
	PotType         string
}

// NewHandState creates a new empty hand state.
func NewHandState() HandState {
	return HandState{
		Players: make(map[string]*PlayerHandState),
		Pots:    []*PotState{{PotType: "main"}},
	}
}

// Exists returns true if the hand has been dealt.
func (s HandState) Exists() bool {
	return s.HandID != ""
}

// IsComplete returns true if the hand is complete.
func (s HandState) IsComplete() bool {
	return s.Status == "complete"
}

// ActivePlayerCount returns number of players still in the hand.
func (s HandState) ActivePlayerCount() int {
	count := 0
	for _, p := range s.Players {
		if !p.HasFolded {
			count++
		}
	}
	return count
}

// GetPlayerByPosition finds a player by seat position.
func (s HandState) GetPlayerByPosition(pos int32) *PlayerHandState {
	for _, p := range s.Players {
		if p.Position == pos {
			return p
		}
	}
	return nil
}

// GetPlayerByRoot finds a player by root.
func (s HandState) GetPlayerByRoot(root []byte) *PlayerHandState {
	return s.Players[hex.EncodeToString(root)]
}

// TotalPot returns the total amount in all pots.
func (s HandState) TotalPot() int64 {
	total := int64(0)
	for _, pot := range s.Pots {
		total += pot.Amount
	}
	return total
}

// RebuildState rebuilds hand state from event history.
func RebuildState(eventBook *pb.EventBook) HandState {
	state := NewHandState()

	if eventBook == nil {
		return state
	}

	// Start from snapshot if available
	if eventBook.Snapshot != nil && eventBook.Snapshot.State != nil {
		if eventBook.Snapshot.State.MessageIs(&examples.HandState{}) {
			var snapshot examples.HandState
			if err := eventBook.Snapshot.State.UnmarshalTo(&snapshot); err == nil {
				state = applySnapshot(&snapshot)
			}
		}
	}

	// Apply events since snapshot
	for _, page := range eventBook.Pages {
		if page.Event != nil {
			applyEvent(&state, page.Event)
		}
	}

	return state
}

func applySnapshot(snapshot *examples.HandState) HandState {
	players := make(map[string]*PlayerHandState)
	for _, p := range snapshot.Players {
		key := hex.EncodeToString(p.PlayerRoot)
		players[key] = &PlayerHandState{
			PlayerRoot:    p.PlayerRoot,
			Position:      p.Position,
			HoleCards:     p.HoleCards,
			Stack:         p.Stack,
			BetThisRound:  p.BetThisRound,
			TotalInvested: p.TotalInvested,
			HasActed:      p.HasActed,
			HasFolded:     p.HasFolded,
			IsAllIn:       p.IsAllIn,
		}
	}

	var pots []*PotState
	for _, pot := range snapshot.Pots {
		pots = append(pots, &PotState{
			Amount:          pot.Amount,
			EligiblePlayers: pot.EligiblePlayers,
			PotType:         pot.PotType,
		})
	}
	if len(pots) == 0 {
		pots = []*PotState{{PotType: "main"}}
	}

	return HandState{
		HandID:             snapshot.HandId,
		TableRoot:          snapshot.TableRoot,
		HandNumber:         snapshot.HandNumber,
		GameVariant:        snapshot.GameVariant,
		RemainingDeck:      snapshot.RemainingDeck,
		Players:            players,
		CommunityCards:     snapshot.CommunityCards,
		CurrentPhase:       snapshot.CurrentPhase,
		ActionOnPosition:   snapshot.ActionOnPosition,
		CurrentBet:         snapshot.CurrentBet,
		MinRaise:           snapshot.MinRaise,
		Pots:               pots,
		DealerPosition:     snapshot.DealerPosition,
		SmallBlindPosition: snapshot.SmallBlindPosition,
		BigBlindPosition:   snapshot.BigBlindPosition,
		Status:             snapshot.Status,
	}
}

func applyEvent(state *HandState, eventAny *anypb.Any) {
	typeURL := eventAny.TypeUrl

	switch {
	case strings.HasSuffix(typeURL, ".CardsDealt"):
		var event examples.CardsDealt
		if err := proto.Unmarshal(eventAny.Value, &event); err == nil {
			state.HandID = hex.EncodeToString(state.TableRoot) + "_" + string(rune(event.HandNumber))
			state.TableRoot = event.TableRoot
			state.HandNumber = event.HandNumber
			state.GameVariant = event.GameVariant
			state.DealerPosition = event.DealerPosition
			state.RemainingDeck = event.RemainingDeck
			state.CurrentPhase = examples.BettingPhase_PREFLOP
			state.Status = "betting"

			// Initialize players
			for _, p := range event.Players {
				key := hex.EncodeToString(p.PlayerRoot)
				state.Players[key] = &PlayerHandState{
					PlayerRoot: p.PlayerRoot,
					Position:   p.Position,
					Stack:      p.Stack,
				}
			}

			// Apply hole cards
			for _, pc := range event.PlayerCards {
				key := hex.EncodeToString(pc.PlayerRoot)
				if player := state.Players[key]; player != nil {
					player.HoleCards = pc.Cards
				}
			}
		}

	case strings.HasSuffix(typeURL, "BlindPosted"):
		var event examples.BlindPosted
		if err := proto.Unmarshal(eventAny.Value, &event); err == nil {
			key := hex.EncodeToString(event.PlayerRoot)
			if player := state.Players[key]; player != nil {
				player.Stack = event.PlayerStack
				player.BetThisRound += event.Amount
				player.TotalInvested += event.Amount
			}
			state.Pots[0].Amount = event.PotTotal
			if event.Amount > state.CurrentBet {
				state.CurrentBet = event.Amount
			}
		}

	case strings.HasSuffix(typeURL, "ActionTaken"):
		var event examples.ActionTaken
		if err := proto.Unmarshal(eventAny.Value, &event); err == nil {
			key := hex.EncodeToString(event.PlayerRoot)
			if player := state.Players[key]; player != nil {
				player.Stack = event.PlayerStack
				player.HasActed = true

				switch event.Action {
				case examples.ActionType_FOLD:
					player.HasFolded = true
				case examples.ActionType_ALL_IN:
					player.IsAllIn = true
					player.BetThisRound += event.Amount
					player.TotalInvested += event.Amount
				case examples.ActionType_BET, examples.ActionType_RAISE, examples.ActionType_CALL:
					player.BetThisRound += event.Amount
					player.TotalInvested += event.Amount
				}
			}
			state.Pots[0].Amount = event.PotTotal
			state.CurrentBet = event.AmountToCall
		}

	case strings.HasSuffix(typeURL, "BettingRoundComplete"):
		var event examples.BettingRoundComplete
		if err := proto.Unmarshal(eventAny.Value, &event); err == nil {
			// Reset for next round
			for _, p := range state.Players {
				p.BetThisRound = 0
				p.HasActed = false
			}
			state.CurrentBet = 0

			// Update stacks from snapshot
			for _, snap := range event.Stacks {
				key := hex.EncodeToString(snap.PlayerRoot)
				if player := state.Players[key]; player != nil {
					player.Stack = snap.Stack
					player.IsAllIn = snap.IsAllIn
					player.HasFolded = snap.HasFolded
				}
			}
		}

	case strings.HasSuffix(typeURL, "CommunityCardsDealt"):
		var event examples.CommunityCardsDealt
		if err := proto.Unmarshal(eventAny.Value, &event); err == nil {
			state.CommunityCards = event.AllCommunityCards
			state.CurrentPhase = event.Phase
		}

	case strings.HasSuffix(typeURL, "DrawCompleted"):
		var event examples.DrawCompleted
		if err := proto.Unmarshal(eventAny.Value, &event); err == nil {
			key := hex.EncodeToString(event.PlayerRoot)
			if player := state.Players[key]; player != nil {
				// Replace discarded cards with new cards
				// The event contains the new cards; we need to update hole cards
				// For simplicity, just append new cards (in real impl would replace specific indices)
				if len(event.NewCards) > 0 {
					// Create new hole cards array keeping non-discarded and adding new
					player.HoleCards = append(player.HoleCards[:len(player.HoleCards)-int(event.CardsDiscarded)], event.NewCards...)
				}
			}
			// Update remaining deck
			if int(event.CardsDrawn) <= len(state.RemainingDeck) {
				state.RemainingDeck = state.RemainingDeck[event.CardsDrawn:]
			}
		}

	case strings.HasSuffix(typeURL, "ShowdownStarted"):
		state.Status = "showdown"

	case strings.HasSuffix(typeURL, "CardsRevealed"):
		// Cards revealed during showdown - could store revealed hands

	case strings.HasSuffix(typeURL, "CardsMucked"):
		// Player mucked - could mark as mucked

	case strings.HasSuffix(typeURL, "PotAwarded"):
		var event examples.PotAwarded
		if err := proto.Unmarshal(eventAny.Value, &event); err == nil {
			for _, winner := range event.Winners {
				key := hex.EncodeToString(winner.PlayerRoot)
				if player := state.Players[key]; player != nil {
					player.Stack += winner.Amount
				}
			}
		}

	case strings.HasSuffix(typeURL, "HandComplete"):
		var event examples.HandComplete
		if err := proto.Unmarshal(eventAny.Value, &event); err == nil {
			state.Status = "complete"
			// Update final stacks
			for _, snap := range event.FinalStacks {
				key := hex.EncodeToString(snap.PlayerRoot)
				if player := state.Players[key]; player != nil {
					player.Stack = snap.Stack
				}
			}
		}
	}
}

File: examples/rust/hand/agg/src/state.rs
//! Hand aggregate state.

use std::collections::HashMap;
use std::sync::LazyLock;

use angzarr_client::proto::event_page::Payload;
use angzarr_client::proto::examples::{
    ActionTaken, ActionType, BettingPhase, BettingRoundComplete, BlindPosted, Card, CardsDealt,
    CommunityCardsDealt, DrawCompleted, GameVariant, HandComplete, HandState as ProtoHandState,
    PotAwarded, ShowdownStarted,
};
use angzarr_client::proto::EventBook;
use angzarr_client::StateRouter;
use angzarr_client::UnpackAny;

/// Player's state in the hand.
#[derive(Debug, Clone, Default)]
pub struct PlayerHandState {
    pub player_root: Vec<u8>,
    pub position: i32,
    pub hole_cards: Vec<Card>,
    pub stack: i64,
    pub bet_this_round: i64,
    pub total_invested: i64,
    pub has_acted: bool,
    pub has_folded: bool,
    pub is_all_in: bool,
}

/// Pot state.
#[derive(Debug, Clone, Default)]
pub struct PotState {
    pub amount: i64,
    pub eligible_players: Vec<Vec<u8>>,
    pub pot_type: String,
}

/// Hand aggregate state rebuilt from events.
#[derive(Debug, Default, Clone)]
pub struct HandState {
    pub hand_id: String,
    pub table_root: Vec<u8>,
    pub hand_number: i64,
    pub game_variant: GameVariant,

    // Deck state
    pub remaining_deck: Vec<Card>,

    // Player state
    pub players: HashMap<String, PlayerHandState>, // player_root_hex -> state

    // Community cards
    pub community_cards: Vec<Card>,

    // Betting state
    pub current_phase: BettingPhase,
    pub action_on_position: i32,
    pub current_bet: i64,
    pub min_raise: i64,
    pub pots: Vec<PotState>,

    // Positions
    pub dealer_position: i32,
    pub small_blind_position: i32,
    pub big_blind_position: i32,

    pub status: String, // "dealing", "betting", "showdown", "complete"
}

impl HandState {
    /// Check if the hand exists.
    pub fn exists(&self) -> bool {
        !self.hand_id.is_empty()
    }

    /// Check if the hand is complete.
    pub fn is_complete(&self) -> bool {
        self.status == "complete"
    }

    /// Count active (non-folded) players.
    pub fn active_player_count(&self) -> usize {
        self.players.values().filter(|p| !p.has_folded).count()
    }

    /// Get player by root.
    pub fn get_player(&self, player_root: &[u8]) -> Option<&PlayerHandState> {
        let key = hex::encode(player_root);
        self.players.get(&key)
    }

    /// Get mutable player by root.
    pub fn get_player_mut(&mut self, player_root: &[u8]) -> Option<&mut PlayerHandState> {
        let key = hex::encode(player_root);
        self.players.get_mut(&key)
    }

    /// Get total pot amount.
    pub fn total_pot(&self) -> i64 {
        self.pots.iter().map(|p| p.amount).sum()
    }
}

// Event applier functions for StateRouter

fn apply_cards_dealt(state: &mut HandState, event: CardsDealt) {
    state.hand_id = format!("{}_{}", hex::encode(&state.table_root), event.hand_number);
    state.table_root = event.table_root;
    state.hand_number = event.hand_number;
    state.game_variant = GameVariant::try_from(event.game_variant).unwrap_or_default();
    state.dealer_position = event.dealer_position;
    state.remaining_deck = event.remaining_deck;
    state.current_phase = BettingPhase::Preflop;
    state.status = "betting".to_string();

    // Initialize players from PlayerInHand messages
    for p in &event.players {
        let key = hex::encode(&p.player_root);
        state.players.insert(
            key,
            PlayerHandState {
                player_root: p.player_root.clone(),
                position: p.position,
                stack: p.stack,
                ..Default::default()
            },
        );
    }

    // Apply hole cards
    for pc in &event.player_cards {
        let key = hex::encode(&pc.player_root);
        if let Some(player) = state.players.get_mut(&key) {
            player.hole_cards = pc.cards.clone();
        }
    }
}

fn apply_blind_posted(state: &mut HandState, event: BlindPosted) {
    let key = hex::encode(&event.player_root);
    if let Some(player) = state.players.get_mut(&key) {
        player.stack = event.player_stack;
        player.bet_this_round += event.amount;
        player.total_invested += event.amount;
    }
    if let Some(pot) = state.pots.first_mut() {
        pot.amount = event.pot_total;
    }
    if event.amount > state.current_bet {
        state.current_bet = event.amount;
    }
    // Track min_raise as the big blind (highest blind posted)
    if event.amount > state.min_raise {
        state.min_raise = event.amount;
    }
}

fn apply_action_taken(state: &mut HandState, event: ActionTaken) {
    let key = hex::encode(&event.player_root);
    if let Some(player) = state.players.get_mut(&key) {
        player.stack = event.player_stack;
        player.has_acted = true;

        match ActionType::try_from(event.action).unwrap_or_default() {
            ActionType::Fold => {
                player.has_folded = true;
            }
            ActionType::AllIn => {
                player.is_all_in = true;
                player.bet_this_round += event.amount;
                player.total_invested += event.amount;
            }
            ActionType::Bet | ActionType::Raise | ActionType::Call => {
                player.bet_this_round += event.amount;
                player.total_invested += event.amount;
            }
            _ => {}
        }
    }
    if let Some(pot) = state.pots.first_mut() {
        pot.amount = event.pot_total;
    }
    state.current_bet = event.amount_to_call;
}

fn apply_betting_round_complete(state: &mut HandState, event: BettingRoundComplete) {
    // Reset for next round
    for player in state.players.values_mut() {
        player.bet_this_round = 0;
        player.has_acted = false;
    }
    state.current_bet = 0;

    // Update stacks from snapshot
    for snap in &event.stacks {
        let key = hex::encode(&snap.player_root);
        if let Some(player) = state.players.get_mut(&key) {
            player.stack = snap.stack;
            player.is_all_in = snap.is_all_in;
            player.has_folded = snap.has_folded;
        }
    }

    // For Five Card Draw, transition to Draw phase after preflop
    if state.game_variant == GameVariant::FiveCardDraw {
        let completed = BettingPhase::try_from(event.completed_phase).unwrap_or_default();
        if completed == BettingPhase::Preflop {
            state.current_phase = BettingPhase::Draw;
        }
    }
}

fn apply_community_cards_dealt(state: &mut HandState, event: CommunityCardsDealt) {
    // Remove dealt cards from deck
    let cards_dealt = event.cards.len();
    if state.remaining_deck.len() >= cards_dealt {
        state.remaining_deck = state.remaining_deck[cards_dealt..].to_vec();
    }
    state.community_cards = event.all_community_cards;
    state.current_phase = BettingPhase::try_from(event.phase).unwrap_or_default();
    // Reset betting state for new round
    for player in state.players.values_mut() {
        player.bet_this_round = 0;
        player.has_acted = false;
    }
    state.current_bet = 0;
}

fn apply_draw_completed(state: &mut HandState, event: DrawCompleted) {
    // Update player's hole cards
    let key = hex::encode(&event.player_root);
    if let Some(player) = state.players.get_mut(&key) {
        player.hole_cards = event.new_cards;
    }
    // Remove drawn cards from deck
    let cards_drawn = event.cards_drawn as usize;
    if state.remaining_deck.len() >= cards_drawn {
        state.remaining_deck = state.remaining_deck[cards_drawn..].to_vec();
    }
}

fn apply_showdown_started(state: &mut HandState, _event: ShowdownStarted) {
    state.status = "showdown".to_string();
}

fn apply_pot_awarded(state: &mut HandState, event: PotAwarded) {
    for winner in &event.winners {
        let key = hex::encode(&winner.player_root);
        if let Some(player) = state.players.get_mut(&key) {
            player.stack += winner.amount;
        }
    }
}

fn apply_hand_complete(state: &mut HandState, event: HandComplete) {
    state.status = "complete".to_string();
    // Update final stacks
    for snap in &event.final_stacks {
        let key = hex::encode(&snap.player_root);
        if let Some(player) = state.players.get_mut(&key) {
            player.stack = snap.stack;
        }
    }
}

/// Default state factory for StateRouter.
fn new_hand_state() -> HandState {
    HandState {
        pots: vec![PotState {
            pot_type: "main".to_string(),
            ..Default::default()
        }],
        ..Default::default()
    }
}

/// StateRouter for fluent state reconstruction.
///
/// Type names are extracted via reflection using `prost::Name::full_name()`.
static STATE_ROUTER: LazyLock<StateRouter<HandState>> = LazyLock::new(|| {
    StateRouter::with_factory(new_hand_state)
        .on::<CardsDealt>(apply_cards_dealt)
        .on::<BlindPosted>(apply_blind_posted)
        .on::<ActionTaken>(apply_action_taken)
        .on::<BettingRoundComplete>(apply_betting_round_complete)
        .on::<CommunityCardsDealt>(apply_community_cards_dealt)
        .on::<DrawCompleted>(apply_draw_completed)
        .on::<ShowdownStarted>(apply_showdown_started)
        .on::<PotAwarded>(apply_pot_awarded)
        .on::<HandComplete>(apply_hand_complete)
});

/// Rebuild hand state from event history.
pub fn rebuild_state(event_book: &EventBook) -> HandState {
    // Start from snapshot if available
    if let Some(snapshot) = &event_book.snapshot {
        if let Some(snapshot_any) = &snapshot.state {
            if let Ok(proto_state) = snapshot_any.unpack::<ProtoHandState>() {
                let mut state = apply_snapshot(&proto_state);
                // Apply events since snapshot
                for page in &event_book.pages {
                    if let Some(Payload::Event(event)) = &page.payload {
                        STATE_ROUTER.apply_single(&mut state, event);
                    }
                }
                return state;
            }
        }
    }

    STATE_ROUTER.with_event_book(event_book)
}

fn apply_snapshot(snapshot: &ProtoHandState) -> HandState {
    let mut players = HashMap::new();
    for p in &snapshot.players {
        let key = hex::encode(&p.player_root);
        players.insert(
            key,
            PlayerHandState {
                player_root: p.player_root.clone(),
                position: p.position,
                hole_cards: p.hole_cards.clone(),
                stack: p.stack,
                bet_this_round: p.bet_this_round,
                total_invested: p.total_invested,
                has_acted: p.has_acted,
                has_folded: p.has_folded,
                is_all_in: p.is_all_in,
            },
        );
    }

    let pots = snapshot
        .pots
        .iter()
        .map(|pot| PotState {
            amount: pot.amount,
            eligible_players: pot.eligible_players.clone(),
            pot_type: pot.pot_type.clone(),
        })
        .collect();

    HandState {
        hand_id: snapshot.hand_id.clone(),
        table_root: snapshot.table_root.clone(),
        hand_number: snapshot.hand_number,
        game_variant: GameVariant::try_from(snapshot.game_variant).unwrap_or_default(),
        remaining_deck: snapshot.remaining_deck.clone(),
        players,
        community_cards: snapshot.community_cards.clone(),
        current_phase: BettingPhase::try_from(snapshot.current_phase).unwrap_or_default(),
        action_on_position: snapshot.action_on_position,
        current_bet: snapshot.current_bet,
        min_raise: snapshot.min_raise,
        pots,
        dealer_position: snapshot.dealer_position,
        small_blind_position: snapshot.small_blind_position,
        big_blind_position: snapshot.big_blind_position,
        status: snapshot.status.clone(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use angzarr_client::pack_event;
    use angzarr_client::proto::event_page;

    #[test]
    fn test_community_cards_dealt_applies_correctly() {
        // Create a CommunityCardsDealt event
        let event = CommunityCardsDealt {
            cards: vec![
                Card { suit: 0, rank: 10 },
                Card { suit: 1, rank: 11 },
                Card { suit: 2, rank: 12 },
            ],
            phase: BettingPhase::Flop as i32,
            all_community_cards: vec![
                Card { suit: 0, rank: 10 },
                Card { suit: 1, rank: 11 },
                Card { suit: 2, rank: 12 },
            ],
            dealt_at: None,
        };

        let event_any = pack_event(&event, "examples.CommunityCardsDealt");

        // Verify STATE_ROUTER applies the event correctly
        let mut state = HandState::default();
        STATE_ROUTER.apply_single(&mut state, &event_any);
        assert_eq!(state.community_cards.len(), 3, "STATE_ROUTER apply failed");
        assert_eq!(state.current_phase, BettingPhase::Flop, "phase not updated");
    }

    #[test]
    fn test_rebuild_from_event_book() {
        use angzarr_client::proto::{EventBook, EventPage, Cover, Uuid};

        // Create CardsDealt event first
        let cards_dealt = CardsDealt {
            table_root: vec![1, 2, 3],
            hand_number: 1,
            game_variant: GameVariant::TexasHoldem as i32,
            dealer_position: 0,
            players: vec![],
            player_cards: vec![],
            remaining_deck: (0..52).map(|i| Card { suit: i / 13, rank: i % 13 }).collect(),
            dealt_at: None,
        };

        // Create CommunityCardsDealt event
        let community = CommunityCardsDealt {
            cards: vec![Card { suit: 0, rank: 10 }],
            phase: BettingPhase::Flop as i32,
            all_community_cards: vec![Card { suit: 0, rank: 10 }],
            dealt_at: None,
        };

        let event_book = EventBook {
            cover: Some(Cover {
                domain: "hand".to_string(),
                root: Some(Uuid { value: vec![1, 2, 3] }),
                ..Default::default()
            }),
            pages: vec![
                EventPage {
                    sequence: 0,
                    payload: Some(event_page::Payload::Event(
                        pack_event(&cards_dealt, "examples.CardsDealt")
                    )),
                    created_at: None,
                },
                EventPage {
                    sequence: 1,
                    payload: Some(event_page::Payload::Event(
                        pack_event(&community, "examples.CommunityCardsDealt")
                    )),
                    created_at: None,
                },
            ],
            snapshot: None,
            next_sequence: 2,
        };

        let state = rebuild_state(&event_book);

        assert_eq!(state.current_phase, BettingPhase::Flop, "phase should be Flop after community dealt");
        assert_eq!(state.community_cards.len(), 1, "should have 1 community card");
    }
}
